### ISSUE TITLE
panic on help when redirecting output and the target halts

### ISSUE THREAD

**Issue by yichengq** [codegangsta/cli#265]
I use the guide example and cli on latest master branch:
```
package main

import (
	"os"

	"github.com/codegangsta/cli"
)

func main() {
	app := cli.NewApp()
	app.Name = "greet"
	app.Usage = "fight the loneliness!"
	app.Action = func(c *cli.Context) {
		println("Hello friend!")
	}

	app.Run(os.Args)
}
```

If you run it in this way, it works fine:
```
$ ./test | grep --peers 
grep: unrecognized option `--peers'
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
	[-e pattern] [-f file] [--binary-files=value] [--color=when]
	[--context[=num]] [--directories=action] [--label] [--line-buffered]
	[--null] [pattern] [file ...]
Hello friend!
```

But when you print help message, it panics:
```
$ ./test -h | grep --peers 
grep: unrecognized option `--peers'
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
	[-e pattern] [-f file] [--binary-files=value] [--color=when]
	[--context[=num]] [--directories=action] [--label] [--line-buffered]
	[--null] [pattern] [file ...]
panic: template: help:1:0: executing "help" at <NAME:
   >: write /dev/stdout: broken pipe

goroutine 1 [running]:
github.com/codegangsta/cli.func·004(0x1a0210, 0x1e7, 0x1428c0, 0x20827eb60)
	/Users/unihorn/gopath/src/github.com/codegangsta/cli/app.go:88 +0x1ef
github.com/codegangsta/cli.ShowAppHelp(0x2082ce000)
	/Users/unihorn/gopath/src/github.com/codegangsta/cli/help.go:100 +0x60
github.com/codegangsta/cli.checkHelp(0x2082ce000, 0x2082aa000)
	/Users/unihorn/gopath/src/github.com/codegangsta/cli/help.go:170 +0x88
github.com/codegangsta/cli.(*App).Run(0x20827eb60, 0x20826a000, 0x2, 0x2, 0x0, 0x0)
	/Users/unihorn/gopath/src/github.com/codegangsta/cli/app.go:134 +0x9e8
main.main()
	/Users/unihorn/gopath/src/github.com/yichengq/test/main.go:17 +0xae
```

from https://github.com/coreos/etcd/issues/3140

**Comment by jszwedko** [comment]
@yichengq thanks for reporting this! It seems to be something specific to the `text/template` code.

**Comment by jszwedko** [comment]
Addressed by #320
